crates/ra_hir_ty/src/display.rs

```diff
@@ -81,36 +133,120 @@ impl<'a, 'b> HirFormatter<'a, 'b> {
     }
 }
 
-pub struct HirDisplayWrapper<'a, T>(&'a dyn HirDatabase, &'a T, Option<usize>, bool);
+pub trait RenderTarget: Sized + Copy {
+    type Error: From<fmt::Error>;
+    fn render_adt(&self, f: &mut HirFormatter<Self>, def_id: AdtId) -> Result<(), Self::Error>;
+}
+
+/// Render types for inlays, doc popups, autocompletion, etc...
+/// Showing `{unknown}` or not qualifying paths is fine here.
+/// There's no reason for this to fail.
+#[derive(Clone, Copy)]
+pub struct RenderDiagnostics;
+
+impl RenderTarget for RenderDiagnostics {
+    type Error = fmt::Error;
+
+    fn render_adt(&self, f: &mut HirFormatter<Self>, def_id: AdtId) -> Result<(), Self::Error> {
+        let name = match def_id {
+            AdtId::StructId(it) => f.db.struct_data(it).name.clone(),
+            AdtId::UnionId(it) => f.db.union_data(it).name.clone(),
+            AdtId::EnumId(it) => f.db.enum_data(it).name.clone(),
+        };
+        write!(f, "{}", name)?;
+        Ok(())
+    }
+}
+
+/// Render types for inserting them in source files.
+/// The generated code should compile, so paths need to be qualified.
+#[derive(Clone, Copy)]
+struct RenderSourceCode {
+    /// The module to render in.
+    /// Required for path qualification.
+    in_module: ModuleId,
+}
+
+#[derive(Debug)]
+pub enum RenderSourceCodeError {
+    UnresolvedPath,
+}
+
+/// Possible errors during formatting
+enum FormatSourceCodeError {
+    /// `RenderSourceCodeError` contains errors we actually want to handle
+    RenderSourceCodeError(RenderSourceCodeError),
+    /// `FmtError` is only required because `HirDisplay::hir_fmt` can return `fmt::Error`
+    FmtError,
+}
+impl From<fmt::Error> for FormatSourceCodeError {
+    fn from(_: fmt::Error) -> Self {
+        Self::FmtError
+    }
+}
+
+impl RenderTarget for RenderSourceCode {
+    /// If the type was not displayable or a path could not be qualified,
+    /// we just return an empty error.
+    type Error = FormatSourceCodeError;
+    fn render_adt(&self, f: &mut HirFormatter<Self>, def_id: AdtId) -> Result<(), Self::Error> {
+        Ok(
+            if let Some(path) =
+                find_path::find_path(f.db.upcast(), ItemInNs::Types(def_id.into()), self.in_module)
+            {
+                write!(f, "{}", path)?;
+            } else {
+                return Err(FormatSourceCodeError::RenderSourceCodeError(
+                    RenderSourceCodeError::UnresolvedPath,
```

Comment:
`Unresolved` is kind of the wrong word, since this is the reverse operation :sweat_smile:

Reply:
`NoPathFound`?

Reply:
Yeah, that seems good.